desc/continuation.py

```diff
@@ -0,0 +1,490 @@
+"""Functions for solving for equilibria with multigrid continuation method."""
+
+import copy
+
+import numpy as np
+
+from desc.equilibrium import EquilibriaFamily, Equilibrium
+from desc.objectives import get_equilibrium_objective, get_fixed_boundary_constraints
+from desc.optimize import Optimizer
+from desc.perturbations import get_deltas
+from desc.utils import Timer
+
+
+def solve_continuation_automatic(  # noqa: C901
+    eq,
+    objective="force",
+    optimizer="lsq-exact",
+    pert_order=2,
+    ftol=1e-2,
+    xtol=1e-4,
+    gtol=1e-6,
+    nfev=100,
+    verbose=1,
+    checkpoint_path=None,
+    **kwargs,
+):
+    """Solve for an equilibrium using an automatic continuation method.
+
+    By default, the method first solves for a no pressure tokamak, then a finite beta
+    tokamak, then a finite beta stellarator. Currently hard coded to take a fixed
+    number of perturbation steps based on conservative estimates and testing. In the
+    future, continuation stepping will be done adaptively.
+
+    Parameters
+    ----------
+    eq : Equilibrium
+        Unsolved Equilibrium with the final desired boundary, profiles, resolution.
+    objective : {"force", "energy", "vacuum"}
+        function to solve for equilibrium solution
+    optimizer : str or Optimzer (optional)
+        optimizer to use
+    pert_order : int
+        order of perturbations to use.
+    ftol, xtol, gtol : float
+        stopping tolerances for subproblem at each step.
+    nfev : int
+        maximum number of function evaluations in each equilibrium subproblem.
+    verbose : integer
+        * 0: no output
+        * 1: summary of each iteration
+        * 2: as above plus timing information
+        * 3: as above plus detailed solver output
+    checkpoint_path : str or path-like
+        file to save checkpoint data (Default value = None)
+
+    Returns
+    -------
+    eqfam : EquilibriaFamily
+        family of equilibria for the intermediate steps, where the last member is the
+        final desired configuration,
+
+    """
+    timer = Timer()
+    timer.start("Total time")
+
+    surface = eq.surface
+    pressure = eq.pressure
+    L, M, N, L_grid, M_grid, N_grid = eq.L, eq.M, eq.N, eq.L_grid, eq.M_grid, eq.N_grid
+    spectral_indexing = eq.spectral_indexing
+
+    mres_step = kwargs.pop("mres_step", 6)
+    pres_step = kwargs.pop("pres_step", 1 / 2)
+    bdry_step = kwargs.pop("bdry_step", 1 / 4)
```

Comment:
These input options should be documented in the docstring.

Reply:
I thought in general we weren't documenting kwargs unless its something the user will likely use? These were mainly for testing and development stuff and could be removed but figured we might as well keep them.

Reply:
I added documentation but @f0uriest should double check it is accurate